go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,541 @@
+		if err != nil {
+			logrus.Errorf("%s Endpount create/update failed", ovnkubeDbEp)
+			return err
+		}
+	}
+
+	// run the cluster controller to init the master
+	err := hacluster.cluster.StartClusterMaster(hacluster.nodeName)
+	if err != nil {
+		return err
+	}
+
+	return hacluster.ovnController.Run()
+}
+
+//updateOvnDbEndpoints Updates the ovnkube-db endpoints. Should be called
+// only if ovnkube-master is leader. This function will create the ovnkube-db endpoints
+// if it doesn't exist.
+func (hacluster *OvnHAMasterController) updateOvnDbEndpoints(ep *kapi.Endpoints, configureInvalidIP bool) error {
+
+	var epIP string
+	if configureInvalidIP {
+		epIP = invalidIPAddress
+	} else {
+		epIP = config.Kubernetes.PodIP
+	}
+
+	epSubsets := []v1.EndpointSubset{
+		{
+			Addresses: []v1.EndpointAddress{
+				{IP: epIP},
+			},
+			Ports: []v1.EndpointPort{
+				{
+					Name: "north",
+					Port: ovnNorthDbPort,
+				},
+				{
+					Name: "south",
+					Port: ovnSouthDbPort,
+				},
+			},
+		},
+	}
+
+	var err error
+	if ep == nil {
+		logrus.Debugf("updateOvnDbEndpoints : Creating the endpoint")
+		// Create the endpoint
+		ovndbEp := v1.Endpoints{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:   config.Kubernetes.OVNConfigNamespace,
+				Name:        ovnkubeDbEp,
+				Annotations: map[string]string{haMasterLeader: hacluster.nodeName},
+			},
+			Subsets: epSubsets,
+		}
+
+		_, err = hacluster.cluster.Kube.CreateEndpoint(config.Kubernetes.OVNConfigNamespace, &ovndbEp)
+		if err != nil {
+			logrus.Errorf("%s Endpount Create failed", ovnkubeDbEp)
+		}
+	} else {
+		logrus.Debugf("updateOvnDbEndpoints : Updating the endpoint")
+		ovndbEp := ep.DeepCopy()
+		ovndbEp.Subsets = epSubsets
+		ovndbEp.Annotations[haMasterLeader] = hacluster.nodeName
+		_, err := hacluster.cluster.Kube.UpdateEndpoint(config.Kubernetes.OVNConfigNamespace, ovndbEp)
+		if err != nil {
+			logrus.Errorf("%s Endpount Update failed", ovnkubeDbEp)
+		}
+	}
+	return err
+}
+
+// ConfigureAsStandby configures the node as standby
+func (hacluster *OvnHAMasterController) ConfigureAsStandby(ep *kapi.Endpoints) error {
+	if !hacluster.manageDBServers {
+		// Nothing to do if not managing db servers.
+		return nil
+	}
+
+	// Get the master ip
+	masterIPList, sbDBPort, nbDBPort, err := extractDbRemotesFromEndpoint(ep)
+	if err != nil {
+		// The db remotes are invalid. Return without doing anything.
+		// Once master updates the endpoints properly we will be notified.
+		logrus.Infof("ConfigureAsStandby : error in extracting DbRemotes From Endpoint")
+		return nil
+	}
+
+	logrus.Infof("ConfigureAsStandby: New leader IP is : [%s]", masterIPList[0])
+
+	// Get the active ovsdb-server address.
+	var stdout string
+	stdout, _, err = util.RunOVNNBAppCtl("ovsdb-server/get-active-ovsdb-server")
+	if err != nil {
+		logrus.Errorf("Getting  active-ovsdb-server of NB ovsdb-server failed")
+		return err
+	}
+
+	activeServerOutOfSync := false
+	s := strings.Split(stdout, ":")
```

Comment:
@numansiddique I think this block could be a separate function, something like:

```
func activeServerOutOfSync(northbound bool, masterIP string, port uint) (bool, error) {
	var stdout, detail string
	var err error

	if northbound {
		detail = "northbound"
		stdout, _, err = util.RunOVNNBAppCtl("ovsdb-server/get-active-ovsdb-server")
	} else {
		detail = "southbound"
		stdout, _, err = util.RunOVNSBAppCtl("ovsdb-server/get-active-ovsdb-server")
	}
	if err != nil {
		logrus.Errorf("Getting  active-ovsdb-server of %s ovsdb-server failed", detail)
		return true, err
	}

	s := strings.Split(stdout, ":")
	if len(s) != 3 {
		return true, nil
	}

	if s[0] != "tcp" || s[1] != masterIP || s[2] != strconv.Itoa(int(port)) {
		return true, nil
	}

	return false, nil
}
```

and then the function gets used in both places?

Reply:
Ack. Done